zerver/views/documentation.py

```diff
@@ -75,28 +75,52 @@ def get_path(self, article: str) -> Tuple[str, int]:
             http_status = 404
 
         path = self.path_template % (article,)
+        # The following is a somewhat hacky approach to extract titles from articles.
+        # Hack: `context["article"] has a leading `/`, so we use + to add directories.
+        article_path = os.path.join(settings.DEPLOY_ROOT, "templates") + path
+        endpoint_name = ""
+        endpoint_method = ""
+        if (not os.path.exists(article_path)) and self.path_template == "/zerver/api/%s.md":
+            endpoint_path = article.replace("-", "_")
+            try:
+                endpoint_name, endpoint_method = get_endpoint_from_operationid(endpoint_path)
+                path = "/zerver/api/api-doc-template.md"
+            except AssertionError:
+                return (self.path_template % ("missing",), 404, "", "")
         try:
             loader.get_template(path)
-            return (path, http_status)
+            return (path, http_status, endpoint_name, endpoint_method)
         except loader.TemplateDoesNotExist:
-            return (self.path_template % ("missing",), 404)
+            return (self.path_template % ("missing",), 404, endpoint_name, endpoint_method)
 
     def get_context_data(self, **kwargs: Any) -> Dict[str, Any]:
         article = kwargs["article"]
         context: Dict[str, Any] = super().get_context_data()
-        (context["article"], http_status_ignored) = self.get_path(article)
+        (context["article"], article_http_status, endpoint_name, endpoint_method) = self.get_path(
+            article
+        )
 
         # For disabling the "Back to home" on the homepage
         context["not_index_page"] = not context["article"].endswith("/index.md")
         if self.path_template == "/zerver/help/%s.md":
             context["page_is_help_center"] = True
             context["doc_root"] = "/help/"
-            (sidebar_index, http_status_ignored) = self.get_path("include/sidebar_index")
+            (
+                sidebar_index,
+                http_status_ignored,
+                endpoint_name_ignored,
+                endpoint_method_ignored,
+            ) = self.get_path("include/sidebar_index")
             title_base = "Zulip Help Center"
         else:
             context["page_is_api_center"] = True
             context["doc_root"] = "/api/"
-            (sidebar_index, http_status_ignored) = self.get_path("sidebar_index")
+            (
+                sidebar_index,
+                http_status_ignored,
+                endpoint_name_ignored,
+                endpoint_method_ignored,
+            ) = self.get_path("sidebar_index")
```

Comment:
This calling convention is really messy.  Rather than returning N variables, can you define a `@dataclass` for the return value of `self.get_path` in a prep commit (with no functional changes) with the `sidebar_index` and `http_status` fields, and then later commits can extend that object?  

I'm also not sure we need to modify `get_path` in the /help/ code path at all; could we instead just have the /api/ code be in a function only present on that subclass?

Reply:
Changed  to use the dataclass.

Splitting the two functions would've duplicated a lot of code IMO, since the endpoint_name and endpoin_method are coming as a by-product of checking the http_status for the endpoint(to differentiate between an invalid endpoint where we return 404, or a valid endpoint whose file got deleted, in case we return 200).